src/plugins/console/public/application/containers/editor/legacy/console_editor/editor_output.tsx

```diff
@@ -86,15 +87,17 @@ function EditorOutputUI() {
         data
           .map((result) => {
             const { value, contentType } = result.response;
+
+            let editorOutput;
             if (readOnlySettings.tripleQuotes && isJSONContentType(contentType)) {
-              return safeExpandLiteralStrings(value as string);
+              editorOutput = safeExpandLiteralStrings(value as string);
             }
+
             if (isMapboxVectorTile(contentType)) {
-              return i18n.translate('console.outputCannotPreviewBinaryData', {
-                defaultMessage: 'Cannot preview binary data.',
-              });
+              const output = new VectorTile(new Protobuf(value));
+              editorOutput = JSON.stringify(output, null, '\t');
```

Comment:
Instead of displaying the entire vector tile contents, mostly, we will want a more summarized view. I was thinking showing number of features for hits and aggs layer along with properterties for first feature. Also show the properties for the meta layer.

@iverase @thomasneirynck 
What do you want displayed from a vector tile in console?

Reply:
I have no strong opinion here. Depends of the purpose of this interface:

- If the purpose is for debugging, then we should get the full response.
- If the purpose is user consumption, then a summarized view would be nice.

Reply:
wondering if @nyurik has thoughts here, since he developed a similar tool for the desktop.

My main feedback would be that showing `JSON.stringify(new VectorTile())` just shows implementation details of that mapbox/VectorTile class (e.g. includes the raw _pbf buffer etc..), Can we leverage the utility methods on that object (`VectorTile#layers` and `VectorTileLayer#feature())` to display something that maps closer to the layout of an actual tile.

```
// Shows a summary of the mvt tile contents. This is read from the original binary mvt response.
{
 layers: {
        meta: [...array of single geojson features ...],
        hits: [...array of 0-n geojson features ...],
        aggs: [....array of 0-n geojson features ...].
   }
}
```

So it would be less about showing raw contents (which is just binary), but something readable.

So we'd try to aim for some happy middle between debugability and user-consumption, because imho, debugability in a query-editor like dev-console is really limited since the return is binary anyway.

Reply:
Tried to get the same result as in the last example:
https://www.elastic.co/guide/en/elasticsearch/reference/master/search-vector-tile-api.html#search-vector-tile-api-api-example
I removed some raw content to make the response more readable:
<details>
<summary>Click to expand</summary>

```json
{
  "hits": {
    "features": [
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3585,
              "y": 2204
            }
          ]
        },
        "properties": {
          "_id": "9E4G5n8BdR0AZJn3bL4p",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3175,
              "y": 3576
            }
          ]
        },
        "properties": {
          "_id": "Pk4G5n8BdR0AZJn3bL8q",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3585,
              "y": 2204
            }
          ]
        },
        "properties": {
          "_id": "5k4G5n8BdR0AZJn3dd2E",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3175,
              "y": 3576
            }
          ]
        },
        "properties": {
          "_id": "hk4G5n8BdR0AZJn3d-pm",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3585,
              "y": 2204
            }
          ]
        },
        "properties": {
          "_id": "Ok4G5n8BdR0AZJn3eO82",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3585,
              "y": 2204
            }
          ]
        },
        "properties": {
          "_id": "5U4G5n8BdR0AZJn3fv0c",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 2603,
              "y": 1476
            }
          ]
        },
        "properties": {
          "_id": "wk8G5n8BdR0AZJn3gAER",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 2603,
              "y": 1476
            }
          ]
        },
        "properties": {
          "_id": "Lk8G5n8BdR0AZJn3gglJ",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3585,
              "y": 2204
            }
          ]
        },
        "properties": {
          "_id": "Zk8G5n8BdR0AZJn3hQ8N",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3585,
              "y": 2204
            }
          ]
        },
        "properties": {
          "_id": "_U8G5n8BdR0AZJn3hQ8N",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3175,
              "y": 3576
            }
          ]
        },
        "properties": {
          "_id": "tk8G5n8BdR0AZJn3ihdE",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3585,
              "y": 2204
            }
          ]
        },
        "properties": {
          "_id": "Sk8G5n8BdR0AZJn3ihhF",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3585,
              "y": 2204
            }
          ]
        },
        "properties": {
          "_id": "q08G5n8BdR0AZJn3kCPq",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 2603,
              "y": 1476
            }
          ]
        },
        "properties": {
          "_id": "q08G5n8BdR0AZJn3kiYl",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 2603,
              "y": 1476
            }
          ]
        },
        "properties": {
          "_id": "4k8G5n8BdR0AZJn3kif2",
          "_index": "kibana_sample_data_logs"
        },
        "type": 1
      }
    ],
    "version": 2,
    "extent": 4096,
    "length": 15
  },
  "aggs": {
    "features": [
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3584,
              "y": 2208
            },
            {
              "x": 3600,
              "y": 2208
            },
            {
              "x": 3600,
              "y": 2192
            },
            {
              "x": 3584,
              "y": 2192
            },
            {
              "x": 3584,
              "y": 2208
            }
          ]
        },
        "properties": {
          "_key": "16/12256/25225",
          "_count": 8
        },
        "type": 3
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 2592,
              "y": 1488
            },
            {
              "x": 2608,
              "y": 1488
            },
            {
              "x": 2608,
              "y": 1472
            },
            {
              "x": 2592,
              "y": 1472
            },
            {
              "x": 2592,
              "y": 1488
            }
          ]
        },
        "properties": {
          "_key": "16/12194/25180",
          "_count": 4
        },
        "type": 3
      },
      {
        "geometry": {
          "coordinates": [
            {
              "x": 3168,
              "y": 3584
            },
            {
              "x": 3184,
              "y": 3584
            },
            {
              "x": 3184,
              "y": 3568
            },
            {
              "x": 3168,
              "y": 3568
            },
            {
              "x": 3168,
              "y": 3584
            }
          ]
        },
        "properties": {
          "_key": "16/12230/25311",
          "_count": 3
        },
        "type": 3
      }
    ],
    "version": 2,
    "extent": 4096,
    "length": 3
  },
  "meta": {
    "features": [
      {
        "geometry": {
          "coordinates": [
            {
              "x": 0,
              "y": 4096
            },
            {
              "x": 4096,
              "y": 4096
            },
            {
              "x": 4096,
              "y": 0
            },
            {
              "x": 0,
              "y": 0
            },
            {
              "x": 0,
              "y": 4096
            }
          ]
        },
        "properties": {
          "_shards.failed": 0,
          "_shards.skipped": 0,
          "_shards.successful": 1,
          "_shards.total": 1,
          "aggregations._count.avg": 5,
          "aggregations._count.count": 3,
          "aggregations._count.max": 8,
          "aggregations._count.min": 3,
          "aggregations._count.sum": 15,
          "hits.total.relation": "eq",
          "hits.total.value": 15,
          "timed_out": false,
          "took": 4
        },
        "type": 3
      }
    ],
    "version": 2,
    "extent": 4096,
    "length": 1
  }
}
```
</details>

Reply:
Looks great. A few suggestions:

- remove the nested `features` property. It's extra indentation I don't think we need. Just make the layers arrays.
- remove the `type` property. It duplicates the geometry-type information (e.g. `1` is a point, `3` is a polygon).
- move the `meta` layer to the top, since it's the most important one, and might get lost under the clutter.

Reply:
@thomasneirynck Here's the updated format:
<details>
<summary>Click to expand</summary>

```json
{
  "meta":[
    {
      "geometry":{
        "coordinates":[
          {
            "x":0,
            "y":4096
          },
          {
            "x":4096,
            "y":4096
          },
          {
            "x":4096,
            "y":0
          },
          {
            "x":0,
            "y":0
          },
          {
            "x":0,
            "y":4096
          }
        ]
      },
      "properties":{
        "_shards.failed":0,
        "_shards.skipped":0,
        "_shards.successful":1,
        "_shards.total":1,
        "aggregations._count.avg":5,
        "aggregations._count.count":3,
        "aggregations._count.max":8,
        "aggregations._count.min":3,
        "aggregations._count.sum":15,
        "hits.total.relation":"eq",
        "hits.total.value":15,
        "timed_out":false,
        "took":1
      }
    }
  ],
  "hits":[
    {
      "geometry":{
        "coordinates":[
          {
            "x":3585,
            "y":2204
          }
        ]
      },
      "properties":{
        "_id":"Ot1w838BsYAjOiKKbfES",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":3175,
            "y":3576
          }
        ]
      },
      "properties":{
        "_id":"hN1w838BsYAjOiKKbfES",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":3585,
            "y":2204
          }
        ]
      },
      "properties":{
        "_id":"XN5w838BsYAjOiKKdAiR",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":3175,
            "y":3576
          }
        ]
      },
      "properties":{
        "_id":"_N5w838BsYAjOiKKdxRD",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":3585,
            "y":2204
          }
        ]
      },
      "properties":{
        "_id":"vN5w838BsYAjOiKKeBdP",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":3585,
            "y":2204
          }
        ]
      },
      "properties":{
        "_id":"W95w838BsYAjOiKKeyiu",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":2603,
            "y":1476
          }
        ]
      },
      "properties":{
        "_id":"RN5w838BsYAjOiKKfypW",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":2603,
            "y":1476
          }
        ]
      },
      "properties":{
        "_id":"vN5w838BsYAjOiKKgC_m",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":3585,
            "y":2204
          }
        ]
      },
      "properties":{
        "_id":"9N5w838BsYAjOiKKhTUA",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":3585,
            "y":2204
          }
        ]
      },
      "properties":{
        "_id":"i95w838BsYAjOiKKhTYA",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":3175,
            "y":3576
          }
        ]
      },
      "properties":{
        "_id":"5d5w838BsYAjOiKKjENV",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":3585,
            "y":2204
          }
        ]
      },
      "properties":{
        "_id":"ed5w838BsYAjOiKKjERV",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":3585,
            "y":2204
          }
        ]
      },
      "properties":{
        "_id":"_d5w838BsYAjOiKKl1PW",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":2603,
            "y":1476
          }
        ]
      },
      "properties":{
        "_id":"_d5w838BsYAjOiKKmVYj",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":2603,
            "y":1476
          }
        ]
      },
      "properties":{
        "_id":"NN5w838BsYAjOiKKmlgf",
        "_index":"kibana_sample_data_logs"
      }
    }
  ],
  "aggs":[
    {
      "geometry":{
        "coordinates":[
          {
            "x":3584,
            "y":2208
          },
          {
            "x":3600,
            "y":2208
          },
          {
            "x":3600,
            "y":2192
          },
          {
            "x":3584,
            "y":2192
          },
          {
            "x":3584,
            "y":2208
          }
        ]
      },
      "properties":{
        "_key":"16/12256/25225",
        "_count":8
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":2592,
            "y":1488
          },
          {
            "x":2608,
            "y":1488
          },
          {
            "x":2608,
            "y":1472
          },
          {
            "x":2592,
            "y":1472
          },
          {
            "x":2592,
            "y":1488
          }
        ]
      },
      "properties":{
        "_key":"16/12194/25180",
        "_count":4
      }
    },
    {
      "geometry":{
        "coordinates":[
          {
            "x":3168,
            "y":3584
          },
          {
            "x":3184,
            "y":3584
          },
          {
            "x":3184,
            "y":3568
          },
          {
            "x":3168,
            "y":3568
          },
          {
            "x":3168,
            "y":3584
          }
        ]
      },
      "properties":{
        "_key":"16/12230/25311",
        "_count":3
      }
    }
  ]
}
```
</details>

Do we really need to skip the "type" property? In this updated response, this property is gone and there is no other info about the type of geometry.

Reply:
> Do we really need to skip the "type" property? In this updated response, this property is gone and there is no other info about the type of geometry.

Reply:
@thomasneirynck I reviewed the RFC 7946 guidelines and tried to form the response in same way:
<details>
<summary>Click to expand</summary>

```json
{
  "meta":[
    {
      "geometry":{
        "type":"Polygon",
        "coordinates":[
          [
            [
              0,
              4096
            ],
            [
              4096,
              4096
            ],
            [
              4096,
              0
            ],
            [
              0,
              0
            ],
            [
              0,
              4096
            ]
          ]
        ]
      },
      "properties":{
        "_shards.failed":0,
        "_shards.skipped":0,
        "_shards.successful":1,
        "_shards.total":1,
        "aggregations._count.avg":5,
        "aggregations._count.count":3,
        "aggregations._count.max":8,
        "aggregations._count.min":3,
        "aggregations._count.sum":15,
        "hits.total.relation":"eq",
        "hits.total.value":15,
        "timed_out":false,
        "took":67
      }
    }
  ],
  "hits":[
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          3585,
          2204
        ]
      },
      "properties":{
        "_id":"-BbN9X8BEs3pqeJ47-oU",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          3175,
          3576
        ]
      },
      "properties":{
        "_id":"QhbN9X8BEs3pqeJ47-sU",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          3585,
          2204
        ]
      },
      "properties":{
        "_id":"GhfN9X8BEs3pqeJ49QJE",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          3175,
          3576
        ]
      },
      "properties":{
        "_id":"uhfN9X8BEs3pqeJ49w44",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          3585,
          2204
        ]
      },
      "properties":{
        "_id":"bhfN9X8BEs3pqeJ49xP6",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          3585,
          2204
        ]
      },
      "properties":{
        "_id":"XhfN9X8BEs3pqeJ4-iNJ",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          2603,
          1476
        ]
      },
      "properties":{
        "_id":"LxfN9X8BEs3pqeJ4-ykP",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          2603,
          1476
        ]
      },
      "properties":{
        "_id":"mxfN9X8BEs3pqeJ4-zDa",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          3585,
          2204
        ]
      },
      "properties":{
        "_id":"dhfN9X8BEs3pqeJ4_zlP",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          3585,
          2204
        ]
      },
      "properties":{
        "_id":"DRfN9X8BEs3pqeJ4_zpQ",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          3175,
          3576
        ]
      },
      "properties":{
        "_id":"ZxfO9X8BEs3pqeJ4Akf1",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          3585,
          2204
        ]
      },
      "properties":{
        "_id":"-xfO9X8BEs3pqeJ4Akf1",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          3585,
          2204
        ]
      },
      "properties":{
        "_id":"fxfO9X8BEs3pqeJ4ClfI",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          2603,
          1476
        ]
      },
      "properties":{
        "_id":"fxfO9X8BEs3pqeJ4C1rB",
        "_index":"kibana_sample_data_logs"
      }
    },
    {
      "geometry":{
        "type":"Point",
        "coordinates":[
          2603,
          1476
        ]
      },
      "properties":{
        "_id":"thfO9X8BEs3pqeJ4DFvc",
        "_index":"kibana_sample_data_logs"
      }
    }
  ],
  "aggs":[
    {
      "geometry":{
        "type":"Polygon",
        "coordinates":[
          [
            [
              3584,
              2208
            ],
            [
              3600,
              2208
            ],
            [
              3600,
              2192
            ],
            [
              3584,
              2192
            ],
            [
              3584,
              2208
            ]
          ]
        ]
      },
      "properties":{
        "_key":"16/12256/25225",
        "_count":8
      }
    },
    {
      "geometry":{
        "type":"Polygon",
        "coordinates":[
          [
            [
              2592,
              1488
            ],
            [
              2608,
              1488
            ],
            [
              2608,
              1472
            ],
            [
              2592,
              1472
            ],
            [
              2592,
              1488
            ]
          ]
        ]
      },
      "properties":{
        "_key":"16/12194/25180",
        "_count":4
      }
    },
    {
      "geometry":{
        "type":"Polygon",
        "coordinates":[
          [
            [
              3168,
              3584
            ],
            [
              3184,
              3584
            ],
            [
              3184,
              3568
            ],
            [
              3168,
              3568
            ],
            [
              3168,
              3584
            ]
          ]
        ]
      },
      "properties":{
        "_key":"16/12230/25311",
        "_count":3
      }
    }
  ]
}
```
</details>

According the guidelines, "type" property must be one of the seven geometry types, but mapbox has only 3 - 'Point', 'LineString', 'Polygon'. Only these cases may have been covered for now.

Reply:
For what is worth, the current output format seems enough to me, given that the features are not in latitude/longitude coordinates, so they cannot be used directly outside the console. I'm using this two steps procedure now to convert a ES `_mvt` search into a valid GeoJSON file using [`vt2geojson`](https://www.npmjs.com/package/@mapbox/vt2geojson):

```
curl -H "Content-Type: application/json" --data @request.json \
 http://elastic:changeme@localhost:9200/madrid_cadastre/_mvt/geom/18/128373/98843 > /tmp/response.pbf

vt2geojson -z 18 -x 128373 -y 98843 /tmp/response.pbf > /tmp/response.geo.json
```

Apart from that, it would be nice if the output had JSON format and folding capabilities like other ES query results, but not a big deal, at least for now.

Reply:
@jsanz thanks for pointing this out, I have added json formatting support for this type of query.